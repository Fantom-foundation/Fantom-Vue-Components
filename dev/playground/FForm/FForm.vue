<template>
    <section>
        <h1 class="h2">Form</h1>

        <f-form v-model="values" @submit="onFormSubmit" @reset="onFormReset" class="grid">
            <fieldset class="col-8">
                <legend>Test form</legend>
                <div class="fform_fieldset_body">
                    <div class="fform_row">
                        <f-form-input name="input1" label="input 1" validate-on-input :validator="inputValidator">
                            <template #bottom="sProps">
                                <f-message v-show="sProps.showErrorMessage" type="error" role="alert">
                                    Fill input 1
                                </f-message>
                            </template>
                        </f-form-input>
                        <div>
                            <f-form-input v-if="c1 === true" key="fff" name="input2" placeholder="12345">
                                <template #label="labelProps">
                                    <label :for="labelProps.id"><b>&#9742; Phone</b></label>
                                </template>
                            </f-form-input>
                        </div>
                    </div>

                    <div class="fform_row">
                        <f-form-input type="textarea" name="textarea1" label="Textarea" />
                    </div>

                    <div class="fform_row">
                        <f-form-input type="checkbox" name="checkbox1" label="Checkbox 1" />
                        <f-form-input type="checkbox" name="checkbox2" label="Checkbox 2" />
                    </div>

                    <div class="fform_row">
                        <f-form-input type="checkbox" name="c1" label="c1" v-model="c1" />
                        <f-form-input type="checkbox" label="c2" v-model="c2" />
                        <button @click="c2 = true">
                            bla
                        </button>
                    </div>

                    <div class="fform_row">
                        <f-form-input type="checkbox" name="checkbox3" label="Multi checkboxes" :data="testos" />
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="checkbox"
                            label="Multi checkboxes"
                            v-model="c3"
                            :data="{ '10': 'c3 1', '20': 'c3 2' }"
                        />
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="checkbox"
                            name="checkbox4"
                            label="Multi checkboxes 2"
                            :data="[
                                { label: 'Checkbox multi 1', value: '10' },
                                { label: 'Checkbox multi 2', value: '20' },
                            ]"
                        />
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="radio"
                            name="radio1"
                            label="Radio"
                            class="fforminput-column"
                            :data="{ '10': 'Radio 1', '20': 'Radio 2', '30': 'Radio 3' }"
                        />
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="select"
                            name="select1"
                            label="Select"
                            :data="[
                                { label: 'Option 1', value: 1 },
                                { label: 'Option 2', value: 2 },
                                { label: 'Option 3', value: 3 },
                            ]"
                        />
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="dropdownlistbox"
                            name="dropdown1"
                            label="dropdownlistbox"
                            :data="data.dropdown"
                        >
                            <template v-slot:buttonLabel="{ item }">
                                &#9733; <b>{{ item.label }}</b>
                            </template>
                            <template v-slot:item="{ item }">
                                &#9733; <b>{{ item.label }}</b>
                            </template>
                        </f-form-input>
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="listbox"
                            name="listbox1"
                            label="Listbox"
                            :data="[
                                { label: 'item 100', value: '100' },
                                { label: 'item 200', id: 'myid', value: '200' },
                                { label: 'item 300 Lorem ipsum', value: '300' },
                            ]"
                        />
                    </div>

                    <div class="fform_row">
                        <div class="fforminput">
                            <f-option type="checkbox" label="Checkbox 1" v-model="chk" />
                            {{ chk }}
                        </div>
                    </div>
                </div>
            </fieldset>
            <div class="col-4">
                <pre>
c1: {{ c1 }}, c2: {{ c2 }}, c3: {{ c3 }} <br />
{{ values }}
</pre>
            </div>
            <div>
                <f-button type="submit" label="Submit" />
                <f-button type="reset" label="Reset" />
            </div>
        </f-form>
    </section>
</template>

<script>
import FForm from '@/components/FForm/FForm.vue';
import FButton from '@/components/FButton/FButton.vue';

import FFormInput from '@/components/FFormInput/FFormInput.vue';
import FOption from '@/components/FOption/FOption.vue';
import FMessage from '@/components/FMessage/FMessage.vue';

export default {
    name: 'FFormPlayground2',

    components: { FMessage, FOption, FFormInput, FButton, FForm },

    data() {
        return {
            testos: { '10': 'Checkbox multi 1', '20': 'Checkbox multi 2' },
            data: {
                select: [
                    { label: 'Option 1', value: 1 },
                    { label: 'Option 2', value: 2 },
                    { label: 'Option 3', value: 3 },
                ],
                dropdown: [
                    { label: 'item 1', value: '10' },
                    { label: 'item 2', id: 'myid', value: '20' },
                    { label: 'item 3 Lorem ipsum', value: '30' },
                    { label: 'item 4', value: '40' },
                    { label: 'item 5', value: '50' },
                    { label: 'item 6', value: '60' },
                ],
                listbox: [
                    { label: 'item 100', value: '100' },
                    { label: 'item 200', id: 'myid', value: '200' },
                    { label: 'item 300 Lorem ipsum', value: '300' },
                ],
            },
            values: {
                input1: 'default',
                dropdown1: '20',
                checkbox2: true,
                checkbox3: ['10'],
                listbox1: '200',
                radio1: '10',
            },
            c1: false,
            c2: false,
            c3: [],
            chk: false,
            // values: getValues(ffData()),
            fff: false,
            /*
            values: {
                input1: 'value',
                input2: '',
                checkbox1: false,
                checkbox2: true,
                checkboxMulti: [],
                radio1: '10',
                dropdown1: '',
                listbox1: '300',
                select1: '',
            },
*/
        };
    },

    methods: {
        inputValidator(_value) {
            return !!_value.trim();
        },

        updateTestos() {
            // this.testos = { '10': 'Checkbox multi 1', '20': 'ssss' };
            this.testos = { '10': 'BITCH' };
            // this.$set(this.testos, '20', 'ssssssss');
            console.log('!@#');
        },

        /*
        updateData() {
            const { values } = this;

            values.checkbox1 = true;
            values.checkboxMulti = ['10', '20'];
            values.dropdown1 = '40';
            values.listbox1 = '200';

            this.data.listbox[0].label = '123';
        },
        */

        onFormSubmit(_data) {
            this.fff = true;
            console.log(_data.values);
            console.log(_data.lastChangedElem);
            // console.log('onFormSubmit2', this.dt);
            // this.updateData();
        },

        onFormReset(_data) {
            console.log('onFormReset', _data.values);
        },

        onFormChange(_data) {
            console.log('onFormChange', _data);
            console.log(_data.eTarget.name);
        },
    },
};
</script>

<style lang="scss">
@import 'fform';
</style>
