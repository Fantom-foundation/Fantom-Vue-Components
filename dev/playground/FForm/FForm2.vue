<template>
    <section>
        <h1 class="h2">Form</h1>

        <f-input required label="Default" />
        <f-input required is-textarea cols="20" rows="5" label="Default" />

        <f-form v-model="values" @submit="onFormSubmit" @reset="onFormReset" class="grid" v-slot="fprops">
            <fieldset class="col-8">
                <legend>Test form</legend>
                <div class="fform_fieldset_body">
                    <h4>elementStates</h4>
                    <p>{{ fprops.elementStates }}</p>
                    <h4>lastChangedElement</h4>
                    <p>{{ fprops.lastChangedElement }}</p>

                    <div class="fform_row">
                        <f-form-input
                            type="combobox"
                            name="combobox"
                            :__throttle-input-interval="250"
                            label="combobox"
                            :data="data.combobox"
                        />
                        <f-form-input
                            type="combobox"
                            name="combobox2"
                            text-is-value
                            label="combobox - text-is-value"
                            :data="data.combobox2"
                        />
                        <f-form-input
                            type="combobox"
                            name="combobox3"
                            select-mode
                            label="combobox - select-mode"
                            :data="data.combobox3"
                        />
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="combobox"
                            name="combobox4"
                            select-mode
                            label="cbx - select-mode, readonly"
                            readonly
                            :data="data.combobox4"
                        />
                        <f-form-input
                            type="custom-combo-box"
                            name="combobox5"
                            :per-page="15"
                            label="cbx - remote"
                            :data="combobox5Data"
                        >
                            <template v-slot:item="{ item }">
                                <div class="flex juc-space-between">
                                    {{ item.label }} <i class="co-grey-5">{{ item.value }}</i>
                                </div>
                            </template>
                        </f-form-input>
                    </div>

                    <f-form-input name="search" type="f-search-field" __placeholder="Search" />

                    <div class="fform_row">
                        <f-form-input required validate-on-input name="inputreq" label="required input" />
                        <f-form-input name="input1" label="input 1" validate-on-input :validator="inputValidator" />
                        <div>
                            <f-form-input v-if="c1 === true" key="fff" name="input2" placeholder="12345">
                                <template #top="props">
                                    <label :for="props.labeledById"><b>&#9742; Phone</b></label>
                                </template>
                                <!--                                <template #label="labelProps">
                                    <label :for="labelProps.id"><b>&#9742; Phone</b></label>
                                </template>-->
                            </f-form-input>
                        </div>
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="dropdownlistbox"
                            name="dropdown1"
                            label="dropdownlistbox"
                            :select-immediately="true"
                            :data="data.dropdown"
                        >
                            <template v-slot:buttonLabel="{ item }">
                                &#9733; <b>{{ item.label }}</b>
                            </template>
                            <template v-slot:item="{ item }">
                                &#9733; <b>{{ item.label }}</b>
                            </template>
                        </f-form-input>
                    </div>

                    <div class="fform_row">
                        <f-form-input type="textarea" name="textarea1" label="Textarea" />
                    </div>
                    <div class="fform_row">
                        <f-form-input
                            auto-resizable-textarea
                            type="textarea"
                            name="textarea2"
                            label="Textarea - auto resizable"
                        />
                    </div>
                    <div class="fform_row">
                        <f-form-input
                            auto-resizable-textarea
                            type="textarea"
                            name="textarea3"
                            label="Textarea - auto resizable"
                        />
                    </div>

                    <div class="fform_row">
                        <f-form-input type="checkbox" name="checkbox1" label="Checkbox 1" />
                        <f-form-input type="checkbox" name="checkbox2" label="Checkbox 2" />
                    </div>

                    <div class="fform_row">
                        <f-form-input type="checkbox" name="c1" label="c1" v-model="c1" />
                        <f-form-input type="checkbox" label="c2" v-model="c2" />
                        <button @click="c2 = true">
                            bla
                        </button>
                    </div>

                    <div class="fform_row">
                        <f-form-input type="checkboxgroup" name="checkbox3" label="Multi checkboxes" :data="testos" />
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="checkboxgroup"
                            label="Multi checkboxes"
                            v-model="c3"
                            :data="{ '10': 'c3 1', '20': 'c3 2' }"
                        />
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="checkboxgroup"
                            name="checkbox4"
                            label="Multi checkboxes 2"
                            :data="[
                                { label: 'Checkbox multi 1', value: '10' },
                                { label: 'Checkbox multi 2', value: '20' },
                            ]"
                        />
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="radiogroup"
                            name="radio1"
                            label="Radio"
                            column
                            :data="{ '10': 'Radio 1', '20': 'Radio 2', '30': 'Radio 3' }"
                        />
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="select"
                            name="select1"
                            label="Select"
                            :validator="_value => (!_value.trim() ? 'Required' : '')"
                            validate-on-change
                            :data="[
                                { label: '---', value: '' },
                                { label: 'Option 1', value: 1 },
                                { label: 'Option 2', value: 2 },
                                { label: 'Option 3', value: 3 },
                            ]"
                            __data="data.select"
                        />
                    </div>

                    <div class="fform_row">
                        <f-form-input
                            type="listbox"
                            name="listbox1"
                            label="Listbox"
                            :data="[
                                { label: 'item 100', value: '100' },
                                { label: 'item 200', id: 'myid3', value: '200' },
                                { label: 'item 300 Lorem ipsum', value: '300' },
                            ]"
                        />
                    </div>

                    <div class="fform_row">
                        <div class="fforminput">
                            <f-option type="checkbox" label="Checkbox 1" v-model="chk" />
                            {{ chk }}
                        </div>
                    </div>
                </div>
            </fieldset>
            <div class="col-4">
                <pre>
c1: {{ c1 }}, c2: {{ c2 }}, c3: {{ c3 }} <br />
{{ values }}
</pre>
            </div>
            <div>
                {{ fprops.errorMessages }}
            </div>
            <div>
                <f-button
                    type="submit"
                    label="Submit"
                    :disabled="fprops.pendingValidation || fprops.errorMessages.length > 0"
                />
                <f-button
                    type="reset"
                    label="Reset"
                    :disabled="fprops.pendingValidation || fprops.errorMessages.length > 0"
                />
            </div>
        </f-form>
    </section>
</template>

<script>
import FForm from '@/components/FForm/FForm.vue';
import FButton from '@/components/FButton/FButton.vue';

import FFormInput from '@/components/FFormInput/FFormInput.vue';
import FOption from '@/components/FOption/FOption.vue';
import { clone } from '@/utils';
import FInput from '@/components/FInput/FInput.vue';

const comboboxData = [
    {
        value: 2,
        label: 'Tani',
    },
    {
        value: 11,
        label: 'Fonsie',
    },
    {
        value: 5,
        label: 'Slade',
    },
    {
        value: 19,
        label: 'Elladine',
    },
    {
        value: 13,
        label: 'Shandie',
        // label: 'Shandie looooooooooooooooooooooooooooooooooooooooooooooooooong',
    },
    {
        value: 14,
        label: 'Noach',
    },
    {
        value: 9,
        label: 'Shela',
    },
    {
        value: 6,
        label: 'Lura',
    },
    {
        value: 12,
        label: 'Lia',
    },
    {
        value: 15,
        label: 'Frayda',
    },
];

export default {
    name: 'FFormPlayground2',

    components: { FInput, FOption, FFormInput, FButton, FForm },

    data() {
        return {
            testos: { '10': 'Checkbox multi 1', '20': 'Checkbox multi 2' },
            data: {
                /*select: [
                    { label: 'Option 1', value: 1 },
                    { label: 'Option 2', value: 2 },
                    { label: 'Option 3', value: 3 },
                ],*/
                select: [
                    { label: 'item 1', value: '10' },
                    { label: 'item 2', value: '20' },
                    { label: 'item 3 Lorem ipsum', value: '30' },
                    { label: 'item 4', value: '40' },
                    { label: 'item 5', value: '50' },
                    { label: 'item 6', value: '60' },
                ],
                dropdown: [
                    { label: 'item 1', value: '10' },
                    { label: 'item 2', id: 'myid', value: '20' },
                    { label: 'item 3 Lorem ipsum', value: '30' },
                    { label: 'item 4', value: '40' },
                    { label: 'item 5', value: '50' },
                    { label: 'item 6', value: '60' },
                ],
                combobox: clone(comboboxData),
                combobox2: clone(comboboxData),
                combobox3: clone(comboboxData),
                combobox4: clone(comboboxData),
                listbox: [
                    { label: 'item 100', value: '100' },
                    { label: 'item 200', id: 'myid2', value: '200' },
                    { label: 'item 300 Lorem ipsum', value: '300' },
                ],
            },
            combobox5Data: [],
            values: {
                input1: '',
                dropdown1: '',
                checkbox2: false,
                checkbox3: [],
                listbox1: '',
                radio1: '',
                search: '',
                textarea3: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
            },
            /*values: {
                input1: 'default',
                dropdown1: '20',
                checkbox2: true,
                checkbox3: ['10'],
                combobox4: 5,
                combobox5: 3,
                listbox1: '200',
                radio1: '10',
                search: 'ww',
            },*/
            c1: false,
            c2: false,
            c3: [],
            chk: false,
            // values: getValues(ffData()),
            fff: false,
            /*
            values: {
                input1: 'value',
                input2: '',
                checkbox1: false,
                checkbox2: true,
                checkboxMulti: [],
                radio1: '10',
                dropdown1: '',
                listbox1: '300',
                select1: '',
            },
*/
        };
    },

    created() {
        setTimeout(() => {
            console.log('???');
            this.values = {
                input1: 'default',
                dropdown1: '20',
                checkbox2: true,
                checkbox3: ['10'],
                combobox4: 5,
                combobox5: 3,
                listbox1: '200',
                radio1: '30',
                search: 'ww',
            };

            /*Object.assign(this.values, {
                input1: 'default',
                dropdown1: '20',
                checkbox2: true,
                checkbox3: ['10'],
                combobox4: 5,
                combobox5: 3,
                listbox1: '200',
                radio1: '30',
                search: 'ww',
            });*/
        }, 1500);

        setTimeout(() => {
            this.combobox5Data = [
                {
                    value: 1,
                    label: 'Default',
                },
                {
                    value: 2,
                    label: 'Default 2',
                },
                {
                    value: 3,
                    label: 'Default 3',
                },
            ];
        }, 1400);
    },

    methods: {
        inputValidator(_value) {
            return new Promise((_resolve, _reject) =>
                setTimeout(() => {
                    if (Math.random() < 0.5) {
                        _reject('error!!');
                    } else {
                        _resolve(!_value.trim() ? 'Required' : '');
                    }
                }, 2000)
            );
            // return !_value.trim() ? 'Required' : '';
        },

        updateTestos() {
            // this.testos = { '10': 'Checkbox multi 1', '20': 'ssss' };
            this.testos = { '10': 'BITCH' };
            // this.$set(this.testos, '20', 'ssssssss');
            console.log('!@#');
        },

        /*
        updateData() {
            const { values } = this;

            values.checkbox1 = true;
            values.checkboxMulti = ['10', '20'];
            values.dropdown1 = '40';
            values.listbox1 = '200';

            this.data.listbox[0].label = '123';
        },
        */

        onFormSubmit(_data) {
            this.fff = true;
            console.log(_data.values);
            console.log(_data.lastChangedElem);
            // console.log('onFormSubmit2', this.dt);
            // this.updateData();
        },

        onFormReset(_data) {
            console.log('onFormReset', _data.values);
        },

        onFormChange(_data) {
            console.log('onFormChange', _data);
            console.log(_data.eTarget.name);
        },
    },
};
</script>

<style lang="scss">
@use 'fform';
</style>
