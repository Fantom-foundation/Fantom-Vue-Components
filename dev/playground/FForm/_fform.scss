.fform {
    .fl {
        display: flex;
        gap: var(--f-spacer-3);
    }

    fieldset {
        display: flex;
        flex-direction: column;
        background: var(--f-color-grey-1);
        padding: var(--f-spacer-5);
        //box-shadow: var(--f-box-shadow-1);

        legend {
            display: block;
            margin: 0;
            font-weight: bold;
        }
    }

    &_row {
        &:not(:last-child) {
            margin-bottom: var(--f-spacer-5);
        }
    }

    .finput {
        width: 100%;

        &_inputcont {
            width: 100%;
        }

        label {
            display: block;
        }
    }

    .fselect {
        width: 100%;

        .sel {
            width: 100%;

            select {
                width: 100%;
            }
        }

        label {
            display: block;
        }
    }

    .finput {
        //width: 100%;
    }

    &.center-form {
        .form-body {
            max-width: 600px;
            margin: 0 auto;
        }
    }
}
