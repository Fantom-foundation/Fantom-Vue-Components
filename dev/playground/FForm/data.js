import { isObject } from '@/utils';

export function getValues(_formData) {
    const data = _formData;
    const values = {};

    if (!isObject(data)) {
        return {};
    }

    Object.keys(data).forEach(_key => {
        values[_key] = 'default' in data[_key] ? data[_key].default : '';
    });

    return values;
}

export function resetValues(_values, _formData) {
    const defaultValues = getValues(_formData);

    Object.keys(_values).forEach(_key => {
        if (_key in defaultValues) {
            _values[_key] = defaultValues[_key];
        }
    });
}

export function ffData() {
    return {
        input1: {
            default: 'value',
        },
        input2: {},
        checkbox1: {
            default: false,
        },
        checkbox2: {
            default: true,
        },
        checkboxMulti: {
            default: [],
        },
        radio1: {
            default: '10',
        },
        dropdown1: {},
        listbox1: {
            default: '300',
        },
        select1: {},
    };
}
