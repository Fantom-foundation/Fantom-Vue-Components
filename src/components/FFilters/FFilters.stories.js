import FFilters from './FFilters.vue';
import FFormInput from '@/components/FFormInput/FFormInput.vue';

export default {
    title: 'FFilters',
    component: FFilters,
};

export const Default = () => ({
    components: { FFilters, FFormInput },
    template: `
        <div>
            <f-filters @submit="data = { ...$event, event: undefined, form: undefined }">
                <f-form-input type="text" :throttle-input-interval="150" placeholder="Search..." name="search" />
                <f-form-input type="combobox" select-mode label="combobox" name="combobox" :data="[
                    { label: '---', value: '' },
                    { label: 'Option 1', value: '10' },
                    { label: 'Option 2', value: '20' },
                    { label: 'Option 3', value: '30' },
                ]" />
            </f-filters>
            <pre class="fos-4 pat-5">{{ data }}</pre>
        </div>
    `,
    data() {
        return {
            data: {},
        };
    },
});
