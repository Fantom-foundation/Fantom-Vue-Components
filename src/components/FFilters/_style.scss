.ffilters {
    width: auto;
    display: inline-flex;
    flex-wrap: wrap;
    align-items: flex-end;
    gap: var(--f-spacer-1, 3px);
}
