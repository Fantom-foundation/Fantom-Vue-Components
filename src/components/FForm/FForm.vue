<template>
    <form ref="form" method="post" class="fform" novalidate @submit="onSubmit" @reset="onReset">
        <slot v-bind="slotProps"></slot>
    </form>
</template>

<script>
import { eventBusMixin } from '../../mixins/event-bus.js';
import { cloneObject, objectEquals } from '../../utils/index.js';
import { findChildrenByName } from '../../utils/vue-helpers.js';

/**
 * The component is intended to work only with the `FFormInput` components.
 */
export default {
    mixins: [eventBusMixin],

    model: {
        prop: 'formValues',
        event: 'input',
    },

    props: {
        /**
         * Text used before error list on form submit, if any form element error exist.
         * Text is invisible and read by screen readers.
         * `%n` will be replaced by actual number of error messages
         */
        formErrorText: {
            type: String,
            default: '%n form errors:',
        },
        values: {
            type: Object,
            default() {
                return {};
            },
        },
        formValues: {
            type: Object,
            default() {
                return null;
            },
        },
        /** Submit form when an element is changed */
        submitOnChange: {
            type: Boolean,
            default: false,
        },
        /** Call preventDefault() on form submit event */
        cancelSubmit: {
            type: Boolean,
            default: true,
        },
        /** Method `isChanged` returns always `false` */
        noChangeCheck: {
            type: Boolean,
            default: false,
        },
        /** Submit form only if form has been changed */
        submitIfChanged: {
            type: Boolean,
            default: false,
        },
    },

    provide() {
        return {
            elements: this.elements,
            // elements: this.elements,
            elementStates: this.elementStates,
            lastChangedElement: this.lastChangedElement,
        };
    },

    data() {
        return {
            // elements: this.formValues || { ...this.values },
            elements: {
                elements: this.formValues || { ...this.values },
                reset: false,
            },
            elementStates: {},
            errorMessages: [],
            pendingValidation: false,
            lastChangedElement: {
                name: '',
                value: '',
                oldValue: '',
            },
        };
    },

    computed: {
        slotProps() {
            return {
                lastChangedElement: this.lastChangedElement,
                elementStates: this.elementStates,
                pendingValidation: this.pendingValidation,
                errorMessages: this.errorMessages,
            };
        },
    },

    watch: {
        lastChangedElement: {
            handler(_value) {
                this.onElementChange(_value);
            },
            deep: true,
        },

        elementStates: {
            handler(_value) {
                const states = cloneObject(_value);

                this.pendingValidation = this.pendingValidationExists(states);
                this.errorMessages = this.collectErrors(states);
            },
            deep: true,
        },

        formValues(_value, _oldValue) {
            if (_value === _oldValue) {
                return;
            }

            if (_value) {
                this.elements.reset = false;

                this.$nextTick(() => {
                    this.elements.reset = true;
                    this.elements.elements = _value;

                    this.$nextTick(() => {
                        this.refreshInitValues();
                    });
                });
            } else {
                this.refreshInitValues();
            }
        },
    },

    created() {
        /** Initial values */
        this._initValues = {};
    },

    mounted() {
        this.$nextTick(() => {
            this.refreshInitValues();
        });
    },

    methods: {
        /**
         * Submit form.
         */
        submit() {
            let eSubmitBtn = this.$refs.form.querySelector('[type="submit"]');

            if (!eSubmitBtn) {
                eSubmitBtn = this.createSubmitButton();
                this.$refs.form.appendChild(eSubmitBtn);
            }

            eSubmitBtn.click();
        },

        createSubmitButton() {
            const elem = document.createElement('button');

            elem.type = 'submit';
            elem.style.display = 'none';

            return elem;
        },

        /**
         * Reset form.
         */
        reset() {
            this.$refs.form.reset();
        },

        /**
         * Collect validation error messages from elements.
         *
         * @param {object} _elementStates
         * @return {array}
         */
        collectErrors(_elementStates) {
            let errors = [];

            Object.keys(_elementStates).forEach(_key => {
                const elementState = _elementStates[_key];

                if (elementState.invalid) {
                    errors = errors.concat(elementState.errors);
                }
            });

            return errors;
        },

        /**
         * Check if a pending validation exits.
         *
         * @param {object} _elementStates
         * @return {boolean}
         */
        pendingValidationExists(_elementStates) {
            const keys = Object.keys(_elementStates);

            for (let i = 0, len1 = keys.length; i < len1; i++) {
                if (_elementStates[keys[i]].pending) {
                    return true;
                }
            }

            return false;
        },

        /**
         * Check form validity.
         *
         * @return {Boolean}
         */
        async checkValidity() {
            const inputs = findChildrenByName(this.$children, 'f-form-input');
            const validationPromises = [];

            // this.pendingValidation = true;

            // collect promises from validation functions
            inputs.forEach(_input => {
                if (_input.willValidate) {
                    validationPromises.push(_input.validate());
                }
            });

            try {
                // await for all promises to settle
                await Promise.all(validationPromises);

                // this.pendingValidation = false;

                this.errorMessages = this.collectErrors(this.elementStates);

                this.ariaReportErrors(this.errorMessages);

                return this.errorMessages.length === 0;
            } catch (_error) {
                this.pendingValidation = false;
                this.errorMessages.push(_error);
                console.error(_error);
                return false;
            }
        },

        ariaReportErrors(_errors = []) {
            const eventBus = this._eventBus;

            if (_errors.length > 0) {
                eventBus.emit(
                    'aria-alert-replace',
                    `${this.formErrorText.replace('%n', _errors.length)} ${_errors.join(' ')}`
                );
            }
        },

        isChanged() {
            // return this.noChangeCheck ? false : !objectEquals(this._initValues, this.elements);
            return this.noChangeCheck ? false : !objectEquals(this._initValues, this.elements.elements);
        },

        getElements() {
            return cloneObject(this.elements.elements);
        },

        getLastChangedElement() {
            return cloneObject(this.lastChangedElement);
        },

        refreshInitValues() {
            this._initValues = this.getElements();
        },

        onElementChange(_value) {
            this.$emit('element-change', cloneObject(_value));

            if (this.submitOnChange) {
                this.submit();
            }
        },

        /**
         * Triggered on form submit.
         *
         * @param {Event} _event
         */
        async onSubmit(_event) {
            if (this.cancelSubmit) {
                _event.preventDefault();
            }

            if (this.submitIfChanged && !this.isChanged()) {
                return;
            }

            try {
                const valid = await this.checkValidity();

                if (valid) {
                    const payload = {
                        values: this.getElements(),
                        lastChangedElem: this.getLastChangedElement(),
                        event: _event,
                        form: this,
                    };

                    this.$emit('submit', payload);
                } else {
                    _event.preventDefault();
                }
            } catch (_error) {
                _event.preventDefault();
                throw _error;
            }
        },

        /**
         * Triggered when form is about to reset.
         *
         * @param {Event} _event
         */
        onReset(_event) {
            const initValues = cloneObject(this._initValues);
            const { elements } = this;

            _event.preventDefault();

            Object.keys(elements).forEach(_key => {
                if (_key in initValues) {
                    elements[_key] = initValues[_key];
                } else {
                    // TODO: default value according to type
                    elements[_key] = '';
                }
            });

            this.$emit('reset', {
                values: cloneObject(initValues),
                event: _event,
            });
        },
    },
};
</script>

<style lang="scss">
@import 'style';
</style>
