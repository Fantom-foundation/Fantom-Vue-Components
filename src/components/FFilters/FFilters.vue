<template>
    <f-form
        ref="form"
        class="ffilters fform-elemsautowidth"
        submit-on-change
        __submit-if-changed
        no-change-check
        v-bind="$attrs"
        v-on="$listeners"
        @submit="$emit('submit', $event)"
        v-slot="slotProps"
    >
        <slot v-bind="slotProps"></slot>
    </f-form>
</template>

<script>
import FForm from '../FForm/FForm.vue';
import { copyMethods } from '@/utils/vue-helpers.js';

/**
 * FForm with inputs displayed side by side
 */
export default {
    name: 'FFilters',

    components: { FForm },

    model: {
        prop: 'formValues',
        event: 'input',
    },

    methods: {
        ...copyMethods(FForm, ['getLastChangedElement'], 'form'),

        getValues() {
            return this.$refs.form.getElements();
        },
    },
};
</script>

<style lang="scss">
@use "style";
</style>
