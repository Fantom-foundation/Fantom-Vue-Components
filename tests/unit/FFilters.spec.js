import { mount } from '@vue/test-utils';
import FFilters from '@/components/FFilters/FFilters.vue';
import FForm from '@/components/FForm/FForm.vue';

let wrapper = null;

function destroyWrapper() {
    if (wrapper) {
        wrapper.destroy();
        wrapper = null;
    }
}

function createWrapper(options = {}) {
    destroyWrapper();
    wrapper = mount(FFilters, options);
}

beforeEach(() => {
    createWrapper();
});

afterEach(() => {
    destroyWrapper();
});

describe('FFilters', () => {
    it('should render correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should pass attributes to the FForm component', () => {
        createWrapper({
            propsData: {
                values: {
                    foo: 'foo',
                },
            },
        });

        const fForm = wrapper.findComponent(FForm);

        expect(fForm.exists()).toBeTruthy();
        expect(fForm.props('values')).toEqual({ foo: 'foo' });
    });

    it('should cointain methods from FForm', () => {
        expect(typeof wrapper.vm.getLastChangedElement).toBe('function');
    });
});
