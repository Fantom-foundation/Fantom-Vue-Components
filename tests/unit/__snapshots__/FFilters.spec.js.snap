// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FFilters should render correctly 1`] = `
<form
  __submit-if-changed=""
  class="fform ffilters fform-elemsautowidth"
  method="post"
  novalidate="novalidate"
/>
`;
